### PR TITLE
amqp/channelpool: forward chanelPoolError on get if there's one

### PR DESCRIPTION
**Pull request by jlevesy**
### What does this PR do?

This PR:
- makes `poolEntity`  internal state thread safe, as it could be written by the supervision goroutine and read concurrently. 
- forwards `poolEntity` internal error on get if there's one. At the moment it's collected but never read. 
- Make `pool.Get` retry indefinitely as soon as the context is valid.

I'm not sure that's our best option here. I'm expecting the caller to retry but I have no clue if it is enforced at the moment.

Relates to https://github.com/upfluence/backlog/issues/663

### Good PR checklist

- [x] Title makes sense
- [x] Is against the correct branch
- [x] Only addresses one issue
- [x] Properly assigned
- [ ] Added/updated tests
- [ ] Added/updated documentation
- [x] Properly labeled

### Additional Notes

Draft for now, I'm not sure of this. But this part is smelly.

### CODE REVIEW

File: amqp/channelpool/pool.go [hunk 1]
```diff
@@ -117,13 +119,20 @@ func (p *poolEntity) supervise(ctx context.Context) {
 			if err != nil {
 				log.WithError(err).Warning("AMQPChannelPool channel closed")
 			}
-
+			p.mu.Lock()
```
Comment (Sypheos) [inline amqp/channelpool/pool.go:122]: wdyt about saving `err` and `closed` is local var then locking after the select

File: amqp/channelpool/pool.go [hunk 2]
```diff
@@ -154,6 +163,14 @@ func (p *pool) Get(ctx context.Context) (*amqp.Channel, error) {
 	}
 
 	pe := e.(*poolEntity)
+
+	pe.mu.RLock()
+	defer pe.mu.RUnlock()
+
+	if pe.err != nil {
+		return nil, err
+	}
```
Comment (Sypheos) [inline amqp/channelpool/pool.go:172]: nit: factorize in a `poolEntity.Error()` method

Comment (jlevesy) [reply amqp/channelpool/pool.go:172]: I also need to handle closed here.

File: amqp/channelpool/pool.go [hunk 3]
```diff
@@ -154,6 +163,14 @@ func (p *pool) Get(ctx context.Context) (*amqp.Channel, error) {
 	}
```
Comment (AlexisMontagne) [inline amqp/channelpool/pool.go:163]: should we here put the get in a for loop and keep on trying to pull out a channel as long as the context is not done?

Comment (jlevesy) [reply amqp/channelpool/pool.go:163]: Yep, I dunno what was expected by the caller, but that's definitely an option. Will do

Comment (jlevesy) [reply amqp/channelpool/pool.go:197]: Any risk to flood the rabbitMQ cluster with channels doing this ? 

If one host nodes goes down, all the node that attempt to consume one of the queues of this host node will spam the other rabbitMQ nodes. 

I would go for a backoff in that case, wdyt ?

**Comment by jlevesy** [comment]
I'll be closing this PR, the issue comes from uamqp.

**Pull request closed without merging**